rolling-shutter/mocksequencer/processor.go

```diff
@@ -1,24 +1,80 @@
 package mocksequencer
 
 import (
+	"context"
 	"math/big"
+	"sort"
+	"sync"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/pkg/errors"
 	"github.com/rs/zerolog/log"
 	txtypes "github.com/shutter-network/txtypes/types"
 )
 
+const (
+	BaseFee  = 22000
+	GasLimit = 2200000000
+)
+
 type blockData struct {
 	baseFee  *big.Int
 	gasLimit uint64
 }
 
+type activationBlockMap[T any] struct {
+	mux sync.RWMutex
+	mp  map[uint64]T
+}
+
+func newActivationBlockMap[T any]() *activationBlockMap[T] {
+	return &activationBlockMap[T]{
+		mp:  map[uint64]T{},
+		mux: sync.RWMutex{},
+	}
+}
+
+func (a *activationBlockMap[T]) Set(val T, block uint64) {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+	a.mp[block] = val
+}
+
+func (a *activationBlockMap[T]) Find(block uint64) (T, error) {
+	var (
+		foundVal T
+		i        int
+	)
+	a.mux.RLock()
+	defer a.mux.RUnlock()
+
+	blocks := make([]uint64, len(a.mp))
+	for k := range a.mp {
+		blocks[i] = k
+		i++
+	}
+	if len(blocks) == 0 {
+		return foundVal, errors.New("no value was set")
+	}
+
+	sort.Slice(blocks, func(i, j int) bool { return blocks[i] < blocks[j] })
+
+	// search for index with next highest activation block
+	idx := sort.Search(len(blocks), func(i int) bool {
+		return blocks[i] > block
```

Comment:
why not search for `blocks[i] >= block` directly?

Reply:
we actually are looking for `i` such that`blocks[i] <= block` AND  `blocks[i+1] > block` but only if there is a higher block. Searching for `blocks[i] <= block` was not possible because of some properties of the search function I believe.

I decided to not try to encode everything in the closure (also because of index out of bounds handling) but rather search the first block that doesn't match the desired condition anymore. The search returns `len(blocks)` when no such index is present.

Reply:
But I found another solution that simplifies the code